src/approval/mod.rs

```diff
@@ -0,0 +1,397 @@
+        testing_env, AccountId, BorshStorageKey,
+    };
+
+    use crate::{approval::ApprovalState, rbac::Rbac};
+    use crate::{near_contract_tools, slot::Slot};
+
+    use super::{Action, Approval};
+
+    #[derive(BorshSerialize, BorshStorageKey)]
+    enum Role {
+        Multisig,
+    }
+
+    #[derive(BorshSerialize, BorshDeserialize, Debug, PartialEq, Eq, Clone, Copy)]
+    enum MyAction {
+        SayHello,
+        SayGoodbye,
+    }
+
+    impl Action for MyAction {
+        type Output = &'static str;
+
+        fn execute(self) -> Self::Output {
+            match self {
+                Self::SayHello => {
+                    println!("Hello!");
+                    "hello"
+                }
+                Self::SayGoodbye => {
+                    println!("Goodbye!");
+                    "goodbye"
+                }
+            }
+        }
+    }
+
+    #[derive(Rbac)]
+    #[rbac(roles = "Role")]
+    #[near_bindgen]
+    struct Contract {}
+
+    #[near_bindgen]
+    impl Contract {
+        #[init]
+        pub fn new(threshold: u8) -> Self {
+            let contract = Self {};
+
+            <Self as Approval<_, _, _>>::init(MultisigConfig { threshold });
+
+            contract
+        }
+    }
+
+    impl Approval<MyAction, MultisigApprovalState, MultisigConfig> for Contract {
+        fn root() -> Slot<()> {
+            Slot::new(b"a")
+        }
+    }
+
+    #[derive(BorshSerialize, BorshDeserialize, Debug)]
+    struct MultisigConfig {
+        pub threshold: u8,
+    }
+
+    #[derive(BorshSerialize, BorshDeserialize, Default, Debug)]
+    struct MultisigApprovalState {
+        pub approved_by: Vec<AccountId>,
+    }
+
+    impl ApprovalState<MultisigConfig> for MultisigApprovalState {
+        fn is_approved(&self, config: &MultisigConfig) -> bool {
+            self.approved_by
+                .iter()
+                .filter(|account| {
+                    // in case a signatory's role was revoked in the meantime
+                    Contract::has_role(&account, &Role::Multisig)
+                })
+                .count()
+                >= config.threshold as usize
+        }
+
+        fn attempt_approval(&mut self, _args: Option<String>, _config: &MultisigConfig) {
+            let predecessor = env::predecessor_account_id();
+            require!(
+                Contract::has_role(&predecessor, &Role::Multisig),
+                "Must have multisig role",
+            );
+            require!(
+                !self.approved_by.contains(&predecessor),
+                "Already approved by this account",
+            );
+
+            self.approved_by.push(predecessor);
+        }
+
+        fn attempt_rejection(&mut self, _args: Option<String>, _config: &MultisigConfig) -> bool {
+            let predecessor = env::predecessor_account_id();
+            require!(
+                Contract::has_role(&predecessor, &Role::Multisig),
+                "Must have multisig role",
```

Comment:
[Recommendation] Use constants for error messages for consistency throughout repo

Reply:
Hence not ready for merging yet :)

Reply:
Actually this is part of a test so it doesn't matter either way